### ISSUE TITLE
Do not discard results of methods marked as pure

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#34098]
In .NET APIs, calling methods for side effects is the norm. Thus, it's generally OK to call a method and discard the result. For example, `List<T>`s `Remove()` method returns a Boolean, indicating whether anything was removed. Ignoring this is just fine.

However, there are cases where ignoring the return value is a 100% a bug. For example, `ImmutableList<T>`'s `Add()` has no side effects and instead returns a new list with the item being added. If you're discarding the return value, then you're not actually doing anything but heating up your CPU and give work to the GC. Fortunately, there is an existing attribute in the BCL, specifically `System.Diagnostics.Contracts.PureAttribute`. Immutable collections already use this attribute to annotate all the APIs that do no have any side effects.

### Proposal

If a method is annotated with `[Pure]`, discarding the return value should be a flagged:

```C#
var x = ImmutableArray<int>.Empty;
x.Add(42); // This should be flagged
```

This should also extend to object initializers if the bound `Add` method is marked as `[Pure]`:

```C#
var x = new ImmutableArray<int>() {
    42 // This should be flagged
}
```

**Comment by terrajobst** [comment]
One could argue that the collection initializer work isn't necessary because types like this will generally not have a constructor. `ImmutableArray<T>` is an exception which is handled by #34096.

**Comment by GrabYourPitchforks** [comment]
Related: [FxCop rule CA1806](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1806) "Do not ignore method results"

**Comment by stephentoub** [comment]
We have a lot of methods with no side-effects that aren't annotated with Pure, which is part of the defunct System.Diagnostics.Contract.  Is the intent of this that we start using it on all methods where it's relevant?  Would we put this on `string.Replace` for example?

**Comment by terrajobst** [comment]
I don't think we need to add it on all methods that are side-effect free, but we should do it on those where mistakes are likely. Once the analyzer is there, we only need to apply the attribute, which is nice. Not sure `string` is likely these days, but for immutable collections it's common when people convert code to use it. Having it there would certainly help.

**Comment by TonyValenti** [comment]
Why not have the C# compiler automatically add [Pure] to methods it detects are pure?  I can see this as a great enhancement that would actually allow the compiler to do further optimizations with common expressions.

**Comment by terrajobst** [comment]
Because it's generally hard to follow this. For example, a method that lazily initializes a cache isn't side effect free (it modifies a field) but it's basically not observable. I think trying to create Haskell out of C# won't fly.

**Comment by TonyValenti** [comment]
Except in that situation, the compiler that compiles the LazyInitializeCache function will know that it isn't pure and wouldn't decorate it as such.

**Comment by stephentoub** [comment]
> we should do it on those where mistakes are likely

I'm fine with having an analyzer + attribute that we can use to mark methods whose return values shouldn't be ignored.  My concern is using [Pure] for that, as a) it wasn't intended for this purpose, b) it's part of a defunct contracts system, c) it's not comprehensive enough, but most importantly d) there are cases where you don't want the return value to be ignored but your method isn't pure (e.g. Stream.Read isn't pure, but ignoring the return value is a recipe for corruption).  Midori had a very robust permissions system, much more expressive than [Pure], yet it still had a [DoNotIgnoreReturnValue] attribute that was separate.  If we want such an analyzer, seems reasonable for it to also respect [Pure] as an aside, but I think it's focus should be on a dedicated attribute for the purpose.  Such an attribute could also be used on out parameters, even ref parameters, which are just another form of return.  We should also think about async methods, where the T in a Task<T> shouldn't be ignored, but having [Pure] on such a method is unlikely to be true, nor is just ensuring that the task itself is consumed, but rather that its returned value is used.

**Comment by terrajobst** [comment]
Works for me

**Comment by dellamonica** [comment]
I think this issue being closed automatically is a mistake. There is real benefit in having an attribute DoNotIgnoreReturnValue and code analysis that warns when the return is ignored. 

(Such a bug just bit me a few hours ago with ImmutableList<T>.Add, and it could have been trivially caught.)

**Comment by tannergooding** [comment]
@stephentoub, how, roughly speaking, would this be different from IDE0058/IDE0059?

Just that it would be "on by default" and would help catch cases like `ImmutableArray` where the original value is not "up to date"?

**Comment by tannergooding** [comment]
Thanks, makes sense!

> lots of work resulting in lots of suppressions to avoid false positives

Well, except this bit. I've never used suppressions here, just used the discard pattern `_ = set.Add(x);` and the IDE quick fix when first enabling it on new codebases.

**Comment by stephentoub** [comment]
> I've never used suppressions here

That's a suppression, just a different form :)

**Comment by terrajobst** [comment]
Agreed. Also, in my opinion it does very little to improve the readability of the code and arguably makes it worse.

**Comment by carlossanlop** [comment]
So from [this comment](https://github.com/dotnet/runtime/issues/34098#issuecomment-604439020), there are 2 work items to tackle with this issue:

Work items:
- Add the new attribute `[DoNotIgnoreReturnValue]`
- Add an analyzer:
  - It should detect methods decorated with the new `[DoNotIgnoreReturnValue]` attribute.
  - It should detect async methods that return `Task`.
  - It does not have a code fixer. Is up to the user to decide what to do with the return value.

The `[Pure]` attribute is ignored.

Suggested category: Reliability
Suggested severity: warning

**Comment by stephentoub** [comment]
>  Is [Conditional("CONTRACTS_FULL")] needed?

No.  It's part of the legacy contracts system.  This should not be included.

> System.Diagnostics.Contracts

It should be in System.Diagnostics.CodeAnalysis.

> It should detect async methods that return Task.

Can you clarify what this is intended to mean?  Consider a method like:
```C#
[DoNotIgnoreReturnValue]
public Task<int> ReadAsync(...);
```
What does the "do not ignore" refer to?  Will it warn for:
```C#
ReadAsync(...);
```
where the returned task is ignored?  Will it warn for:
```C#
await ReadAsync(...);
```
where the `Task<int>` isn't ignored but the `int` extracted from the `Task<int>` and returned from the `await` is ignored?  What about if the API returned `ValueTask<T>` instead of `Task<T>`, or some other awaitable?

**Comment by eerhardt** [comment]
I've updated the top post with the current proposal. Please take a look and provide feedback. Once we are happy with the proposal, I can mark this as `api-ready-for-review`.

**Comment by eerhardt** [comment]
One issue I have with the current design is that I see methods falling into 2 categories:

1. Ignoring the result of this method is basically guaranteed to be a bug in your code - i.e. Reliability. (Stream.Read, ImmutableCollection.Add, `string.Replace/Remove`, etc.)
2. Ignoring the result of this method means the method call is unnecessary - consider removing the method call - i.e. Performance. (LINQ methods, `string.Clone/Copy/Split`, etc)

I am wondering if it would be valuable to add a `bool` parameter to `DoNotIgnoreAttribute`'s ctor, similar to `ObsoleteAttribute`'s `bool error`. This would allow us to keep using `CA1806` Rule ID for the performance cases, and a new Rule ID for the Reliability cases.

**Comment by stephentoub** [comment]
> Ignoring the result of this method means the method call is unnecessary

That _might_ be the case, but it might also just be another case of your code being buggy, i.e. the method call is necessary and the fact that you're ignoring the result means you have a bug.

**Comment by stephentoub** [comment]
> I don't think all of the current CA1806 violations fall into this category

Certainly not "a new object is created but never used", assuming that's not just about "System.Object"... it's not unheard of to create an instance in order to use it for the side-effects of its constructor.  While possibly a perf issue, it's not necessarily a functional one.

But the others, I have trouble differentiating possible reliability impact of ignoring the result of string.Replace (which you've grouped into category 1) and ignoring the result of a LINQ query (which you've grouped into category 2).  I don't know how we'd meaningfully decide for a bool like you suggested whether something should be a warning or an error; we'd have to be able to understand the impact on the code, in any of these cases, and that's not something the analyzer will be able to do.

**Comment by eerhardt** [comment]
> assuming that's not just about "System.Object"...

Looking at [the code](https://github.com/dotnet/roslyn-analyzers/blob/d2ef898a6283858c1224c0114cd87fa27b6cb091/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/DoNotIgnoreMethodResults.cs#L169-L176), it applies to all object creations.

> I don't know how we'd meaningfully decide for a bool like you suggested whether something should be a warning or an error; we'd have to be able to understand the impact on the code, in any of these cases, and that's not something the analyzer will be able to do.

My thinking was that the analyzer wouldn't do this. The API author would decide whether ignoring the result of the method being marked `DoNotIgnore` should be considered a "Reliability" problem or a "Performance" problem.

But if you don't think having this distinction in the attribute makes sense, I think we can drop it.

What do you think about the option to keep all existing methods as `CA1806`, introduce a new Rule ID for `DotNotIgnore`, and don't mark any existing methods (like LINQ and string creation) as `DoNotIgnore`?

**Comment by stephentoub** [comment]
> But if you don't think having this distinction in the attribute makes sense, I think we can drop it.

I think we'd have a really hard time making that call for all of the APIs we produce.

> What do you think about the option to keep all existing methods as CA1806, introduce a new Rule ID for DotNotIgnore, and don't mark any existing methods (like LINQ and string creation) as DoNotIgnore?

@mavasani, do you have an opinion here?  Any sense for how common it is for CA1806 to be used?  It looks like it defaults to Info level.

**Comment by mavasani** [comment]
> I think we'd have a really hard time making that call for all of the APIs we produce.

I personally feel it is going to be valuable to have the new attribute take some parameter (a bool or an enum specifying the issue category if return value is ignored), otherwise we are going to soon run into APIs where it is really bad to ignore its return value, but not significant enough or always guaranteed to lead to a functional bug to break the build. Given the new rule will be build breaking by default (warning and error severities are both considered build breaking), the API author would be unable to mark it with the new attribute. I understand @stephentoub's concern that by adding this additional parameter to the attribute, we are pushing a difficult task to the API authors. It is a tradeoff between additional functionality and additional overhead.

> Any sense for how common it is for CA1806 to be used? It looks like it defaults to Info level.

As you have identified in this issue, [CA1806](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1806) (performance) as well [IDE0058](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0058) (code style/refactoring aid) are half baked rules, based on heuristics/hard coded APIs to flag. Former is too restrictive and misses lot of cases, while latter is too permissive and hence we could not even make it an IDE code style suggestion, it is hidden (refactoring) by default. We have always had lot of requests to be able to mark and flag custom APIs whose return value is ignored, and we ended up adding an [editorconfig option](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#additional-use-results-methods) for users to specify custom APIs to be flagged by the CA1806 analyzer. So the most requested use case for CA1806 is already being served by the new attribute/analyzer being proposed in this issue, which is great.
I feel it is fine to either fold the existing cases handled by CA1806 into the new analyzer rules (if we are planning to add a separate parameter to the attribute to flag non-critical non-build breaking cases) and then deprecate CA1806 OR retain existing cases for CA1806, and have the new analyzer rules only for cases not handled by CA1806.

Note that we do have some precedent here for parallel rules that differ just by the issue category, but have identical analysis: [CA1307: Specify StringComparison for clarity](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1307) and [CA1310: Specify StringComparison for correctness](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1310). CA1307 is disabled by default, while CA1310 is hidden by default (https://github.com/dotnet/roslyn-analyzers/blob/main/src/NetAnalyzers/Core/AnalyzerReleases.Shipped.md).

**Comment by terrajobst** [comment]
Personally, I don't think either category would be sufficient justification for it to be an error/build break.

* Existing code might have a bug, but if it was undetected the world hasn't ended yet. Failing it now seems unnecessary draconian.
* For new code, the user is still writing that code. Failing the build that early might be a productivity hit. Sure, I might not pay attention to warnings because I have already a ton of them, but at least for me the squiggles in the code I just wrote generally get my attention.

**Comment by stephentoub** [comment]
> I don't think either category would be sufficient justification for it to be an error/build break

@eerhardt and I spoke earlier today and this is effectively the same conclusion we reached.  He has more details, but strawman is essentially a default level of info, reliability category, attributing everything where there's a very good chance it indicates a mistake  but not things where there are legitimate reasons you might want to ignore the return value, e.g. there's basically no reason to call string.Replace and ignore the return value, but there are valid reasons you might want to call Enumerable.Single and ignore the return value.

**Comment by GrabYourPitchforks** [comment]
The API is still marked as a suggestion, not ready for review. Would people feel more confident bringing this through review if we spent some time generating a more concrete list of methods to annotate?

I feel the API is fine as is, unless we needed the ability to add a message to it for some reason. e.g., calling Dictionary.TryGetValue and ignoring the out parameter: you should be using ContainsKey instead.

**Comment by carlreinke** [comment]
I set [IDE0058](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0058) to `suggestion` in most of my projects, and it would be nice to have the opposite of `DoNotIgnoreAttribute` — something like `MayIgnoreAttribute`.  The bar for applying it would necessarily be set pretty high — something like "the return value is something you definitely already have".  So, for example, `MayIgnoreAttribute` would be applied to `StringBuilder.Append(...)` but not to `Stream.Seek(...)`.

**Comment by eerhardt** [comment]
> and it would be nice to have the opposite of DoNotIgnoreAttribute — something like MayIgnoreAttribute

I would suggest to open a new issue for that scenario. These two scenarios, while related, aren't dependent on each other so I wouldn't group it into this proposal.

**Comment by eerhardt** [comment]
Marking as "ready for review"

**Comment by bash** [comment]
As a library author, I would love to have this feature.

I see that the proposed attribute can be applied to methods only. Would you consider extending that to allow types too? Applying the attribute to a type would mean that all methods that return that type would be treated as if they were annotated with `[DoNotIgnoreAttribute]`. This would be really useful for functional style result types (`Either` / `Result` / etc.).

(Rust has the very similar `#[must_use]` attribute which can be applied to both types and functions.)

**Comment by bartonjs** [comment]
* We changed Inherited true to Inherited false
* We think that the analyzer should only look at the specific method declaration, and not try to infer the behavior by walking the virtual hierarchy
* For return values, discard assignment (_ = obj.Method()) should count as "not ignoring"
* For out values, discard assignment (out _) should count as "not ignoring".
* The proposed members seem like a good place to start, and once the attribute is merged in send an announcement to have area owners apply it to their areas as they see fit.

```C#
namespace System.Diagnostics.CodeAnalysis
{
    [AttributeUsage(AttributeTargets.ReturnValue | AttributeTargets.Parameter, AllowMultiple = false, Inherited = false)]
    public sealed class DoNotIgnoreAttribute : Attribute
    {
        public DoNotIgnoreAttribute() {}

        public string? Message { get; set; }
    }
}
```

**Comment by eerhardt** [comment]
> Would you consider extending that to allow types too?

We discussed it in the API review and decided to leave this off for now. If it is a common request, we can add it in the future.

**Comment by danmoseley** [comment]
I just made the mistake of `.Append(str)` on a `List<string>` (instead of `.Add(str)`). 

Did we conclude that we would mark something like LINQ's `Append(..)` such that it's flagged by this analyzer?

**Comment by TonyValenti** [comment]
It would be great if any Pure method opted into this as well.

**Comment by eerhardt** [comment]
> Did we conclude that we would mark something like LINQ's Append(..) such that it's flagged by this analyzer?

This is already flagged by `CA1806`. https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1806. In general, all the LINQ methods are not proposed to be marked with `DoNotIgnore` attributes because there are valid cases for calling `.Select` and other methods that can be side-effecting. If there are one-off ones that make sense to be marked, we can consider adding them.

> It would be great if any Pure method opted into this as well.

Pure methods are already covered by `CA1806` as well - https://github.com/dotnet/roslyn-analyzers/blob/e52a3f1c0b45684ddaa041f26a980fa451ea9001/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/DoNotIgnoreMethodResults.cs#L346-L349.

However, no APIs in dotnet/runtime are marked with the Pure attribute anymore. The only ones that were are the System.Collections.Immutable APIs, and those Pure attributes were removed with https://github.com/dotnet/runtime/pull/35118.

**Comment by stephentoub** [comment]
> In general, all the LINQ methods are not proposed to be marked with DoNotIgnore attributes because there are valid cases for calling .Select and other methods that can be side-effecting. If there are one-off ones that make sense to be marked, we can consider adding them.

LINQ methods can generally be split into two categories: those that force enumeration (e.g. Count, Max, Min, Aggregate, etc.) and those that create new lazy enumerables (e.g. Select, Where, Append, Concat, etc.).  We can't attribute the former category as DoNotIgnore, since as you say there are legitimate uses for calling such methods primarily to force evaluation of the query in order to exercise any side effects of those operations.  However, it seems we should be able to annotate methods in the latter category; even if the execution of the query operator has side effects (e.g. invoking the delegate passed to Select), the operator is lazy and won't do any work until the enumerable it returns is enumerated... if that enumerable is ignored, then the operator won't have done anything observable (other than argument validation).

**Comment by carlossanlop** [comment]
Per this comment by @eerhardt, I'm moving this issue to 8.0:

> We discussed it in the API review and decided to leave this off for now. If it is a common request, we can add it in the future.

**Comment by eerhardt** [comment]
> Per this comment by @eerhardt, I'm moving this issue to 8.0:
>> We discussed it in the API review and decided to leave this off for now. If it is a common request, we can add it in the future.

That comment was in response to `Would you consider extending that to allow types too?` in [this comment](https://github.com/dotnet/runtime/issues/34098#issuecomment-1150079218). It didn't apply to this whole feature.

**Comment by Bosch-Eli-Black** [comment]
@carlossanlop Can we move this back to 7.0.0, as per above comment?

**Comment by buyaa-n** [comment]
Estimates:

Analyzer: Large 
Fixer: NA

**Comment by reduckted** [comment]
> `DateTime`,`DateOnly`,`DateTimeOffset`,`TimeSpan`,`TimeOnly` `AddX` methods

Regarding the annotation of date and time methods, I've identified some additional methods in https://github.com/dotnet/roslyn-analyzers/issues/6489 (I created that before I discovered this issue, so it seems redundant now).

`DateTime` and `DateTimeOffset` have `Subtract(...)`, `ToLocalTime()` and `ToUniversalTime()`. 

`TimeSpan` has `Divide(...)`, `Multiply(...)`, `Negate(...)` and `Subtract(...)`.

**Comment by jeffhandley** [comment]
@terrajobst / @eerhardt -- During PR review of the first part of this analyzer (ignoring return values), @stephentoub and I revisited the default level. We propose upgrading the level up to Warning since we're going to limit where the attribute gets applied to be places where it's almost certainly a lurking bug. Any objections?

**Comment by eerhardt** [comment]
No objection from me.

**Comment by jeffhandley** [comment]
As was noted in https://github.com/dotnet/roslyn-analyzers/pull/6546#issuecomment-1546212215, we had some offline discussions about this analyzer and its potential relationship to [IDE0058: Remove unnecessary expression value - .NET | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0058) and [CA1806: Do not ignore method results (code analysis) - .NET | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1806). We have not had the capacity to move that design discussion forward enough to have confidence in landing in a good place during .NET 8.

**Comment by glen-84** [comment]
Could this also allow for specifying a list of fully-qualified return types that must never be ignored?

For example, I have `Result` and `Result<TValue>` types that are returned from many different methods. It would be a lot easier to configure these types as "DoNotIgnore", as opposed to applying the `DoNotIgnore` attribute to every method.

---

See also https://github.com/dotnet/roslyn/issues/47832.

**Comment by glen-84** [comment]
@KalleOlaviNiemitalo I was referring mainly to the namespace. I haven't considered generic type arguments.

**Comment by mykolav** [comment]
Hi,

Sorry for a shameless self-promotion. 

While the official analyzer is in works, you might want to check out this one https://github.com/mykolav/must-use-ret-val-fs  
It emits a diagnostic if it comes across code that ignores the value returned from a method marked with `[MustUseReturnValue]`.

**Comment by timcassell** [comment]
> Would you consider extending that to allow types too?

+1

I wrote a promise library where each promise object is required to either be awaited, or forgotten. Using discard instead of Forget is wrong. It looks like this proposal treats discards as "not ignored", but I would want it to treat is as "ignored". Could that be configurable?

**Comment by julealgon** [comment]
Were there any discussions around alternatives to an attribute in this case? For example, what about using `required` before the return type to signal it must be consumed by the caller?

```csharp
public required Task<int> ReadAsync(...);
```

It feels more "first-class" to me to have it as a language keyword vs an attribute.

Additionally, it could also be used to signal similar semantics for other cases, such as if you want to enforce a given `out` parameter to not be discarded as well:
```csharp
public void MyMethod(required out int value);
```

**Comment by tannergooding** [comment]
We're unlikely to providing anything built-in at this time. We already provide [CA1806: Do not ignore method results](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1806) and this is something that is trivial to enable globally and handle using `discards`.

**Comment by glen-84** [comment]
I don't think CA1806 covers what this proposal is asking for, and I'd ask you to reconsider.

The gap isn't the *generic* case — it's the mechanism and the surface area:

1. **CA1806 has no author-shippable annotation.** Its only extensibility is the consumer-side `dotnet_code_quality.CA1806.additional_use_results_methods` `.editorconfig` list, matched by method name. A library author can't annotate their own return values and have every consumer get the diagnostic transitively across the assembly boundary — every downstream project has to re-declare the method list itself. The whole point of `[DoNotIgnore]` is that the annotation ships *with* the API.

2. **The BCL stays unannotated.** The value here was the curated list — `ImmutableList<T>.Add`, `Stream.Read`, the `DateTime`/`DateOnly`/`TimeSpan` `AddX` methods, `Tuple.Create`, the string-creation APIs. CA1806 out of the box only fires on `new`, string creation, `HRESULT`, LINQ, and `[Pure]`. Item `#1` on the list — the immutable collections — had `[Pure]` removed in #35118, so CA1806 doesn't catch the single case most likely to be a real bug. Closing this means those callsites stay uncovered.

3. **Category.** CA1806 is Performance; this is a correctness/Reliability concern. That distinction was called out in the original proposal (the "Interaction with CA1806" section) as the reason for a separate rule ID — ignoring an `ImmutableList.Add` result isn't "unnecessary work," it's a logic error.

4. **Missing behavior.** The proposal also covered `Task<T>`/`ValueTask<T>` (flagging both the un-awaited task and the discarded awaited result) and `out`/`ref` parameters. CA1806 does neither.

For a method inside your own solution, `additional_use_results_methods` plus a discard (`_ = ...`) as the escape hatch is a reasonable substitute. But the two things that motivated this — an attribute usable across library boundaries, and having the BCL itself annotated — aren't things CA1806 can deliver today.

_Drafted by Claude (Anthropic AI assistant)._

**Comment by tannergooding** [comment]
This is a 6 year old issue with very little activity and it exists in a sea of thousands of other proposals, features, work, bugs, priorities. The simple truth is that the team is unlikely to ever prioritize this in that context and this will just sit open indefinitely, which is not of a benefit to anyone.

Not everything has to be provided in box or by the .NET team, the community can and should build their own analyzers, types, features, and more when there is a need. This should likely even happen first and foremost which helps prove the overall need and usecase, particularly for analyzers which have zero need to be in box.

-----------

I also understand that it is not literally the same as the ask, but it is sufficient enough. Developers shouldn't be arbitrarily discarding results in the first place and should explicitly annotate (such as via the discard pattern) when that is acceptable. They can still have a bug, but they can also have a million other trivial logic bugs as well. Basic tests should cover that, which makes this ask of lower priority and is part of what helps keep it not acted on for 5+ years given every other open request we have (many of which are going to be closed for similar reasons).

**Comment by glen-84** [comment]
That's fair, and I won't argue the prioritization — a 6-year-old request competing with thousands of others is a reasonable thing for the team to decline, and "the community can ship the analyzer and the attribute" is true for the mechanism.

The one piece that doesn't fall to the community: a third-party analyzer can replicate `[DoNotIgnore]` and let downstream projects list their own methods, but it can't annotate the BCL. The half of this ask that was "have `ImmutableList<T>.Add`, `Stream.Read`, the `DateTime.AddX` family, and the string APIs carry the signal in box" is inherently team-only — and item `#1`, the immutable collections, lost even the `[Pure]` coverage in #35118. So a community package proves the demand and covers first-party code, but the in-box surface stays uncovered regardless.

No objection to closing on priority grounds. Leaving that distinction here in case the in-box annotation angle is ever worth revisiting separately from the analyzer.

_Drafted by Claude (Anthropic AI assistant)._

**Comment by tannergooding** [comment]
>  but it can't annotate the BCL

It doesn't have to, you can ship an adjacent text file listing the types/members or build-in a well known list, which several in-box analyzers do already.
